### ISSUE TITLE
How to use ray.tune on cluster node with multiple GPUs

### ISSUE THREAD

**Issue by Shentao-YANG** [ray-project/ray#13240]
Hi,

I would like to use `ray.tune.scheduler` on hyperparameter tuning of Pytorch neural network on one node of the slurm cluster provided by my institution. The node has in total 4 GPUs. Ideally I would like `ray.tune.scheduler` to run and select models parallelly on all 4 GPUs. I tried to adapt the instruction on page **/using-ray-with-pytorch.html** but was not successful --- the `tune.run` function seems only use one GPU (`cuda:0`) .

Could you please provide an example script of using `ray.tune.scheduler` (say, `PBT`, `BOHB`) on hyperparameter tuning of Pytorch model utilizing multiple GPUs?

**Comment by krfricke** [comment]
Most of our examples leverage training on GPUs, e.g. https://docs.ray.io/en/latest/tune/tutorials/tune-pytorch-cifar.html#adding-multi-gpu-support-with-dataparallel

You should pass a `resources_per_trial` argument to `tune.run`:

```
result = tune.run(
    trainable,
    resources_per_trial={"cpu": 1, "gpu": 1}
    # ..) 
```

Please also note that each trial will see the GPU as `cuda:0`, but it will point to different GPUs internally.

If you can provide us with a more thorough example (e.g. your call to `tune.run()`) it is easier for us to find out what is missing.

**Comment by Shentao-YANG** [comment]
Hi, 

Thanks for your reply! 

Here is my `tune.Trainable` object:
`
class Train_Val_VAE(tune.Trainable):
    def setup(self, config):
        print(f"Begin training VAE with config: {config}")
        self.net = VariationalAutoEncoder(latent_dim=config.get("latent_dim", 10),
                                          variational_beta=config.get("variational_beta", 1))
        self.device = "cpu"
        if torch.cuda.is_available():
            self.device = "cuda:0"
            if torch.cuda.device_count() > 1:
                self.net = nn.DataParallel(self.net)
        self.net.to(self.device)
        self.trainloader, self.valloader, _ = \
            get_data_loaders(config.get("data_dir", "./data"), batch_size=int(config.get("batch_size", 128)), val_split=0.2)
        self.optimizer = \
            optim.Adam(params=self.net.parameters(), lr=config.get("lr", 1e-3), weight_decay=config.get("weight_decay", 1e-5))

    def step(self):
        train_VAE(self.net, self.optimizer, self.trainloader, self.device)
        val_loss = test_VAE(self.net, self.valloader, self.device)
        return {"val_loss": val_loss}

    def save_checkpoint(self, checkpoint_dir):
        checkpoint_path = os.path.join(checkpoint_dir, "checkpoint")
        torch.save({
                    "model_state_dict": self.net.state_dict(),
                    "optimizer_state_dict": self.optimizer.state_dict()
                }, checkpoint_path)
        return checkpoint_path

    def load_checkpoint(self, checkpoint_path):
        states = torch.load(checkpoint_path)
        self.net.load_state_dict(states['model_state_dict'])
        self.optimizer.load_state_dict(states['optimizer_state_dict'])
`
Here is my stopper:
`
class CustomStopper(tune.Stopper):
        def __init__(self, max_iter=100, val_loss=(-np.inf)):
            self.should_stop = False
            self.max_iter = max_iter
            self.val_loss = val_loss

        def __call__(self, trial_id, result):
            if not self.should_stop and result["val_loss"] < self.val_loss:
                self.should_stop = True
            return self.should_stop or result["training_iteration"] >= self.max_iter

        def stop_all(self):
            return self.should_stop
`
Here is the main function:
`
def main(num_samples=27, max_num_epochs=200, val_loss=(-np.inf), reduction_factor=3, checkpoint_freq=5,
         cpus_per_trial=(1/1), gpus_per_trial = (1/2)):

    ray.shutdown()
    ray.init(num_cpus=2)

    print(" # check if PytorchTrainble will save/restore correctly before execution")
    validate_save_restore(Train_Val_VAE)
    validate_save_restore(Train_Val_VAE, use_object_store=True)

    data_dir = os.path.abspath("./data")
    get_data_loaders(data_dir, batch_size=64)  # Download data for all trials before starting the run
    config = {
        "data_dir": data_dir,
        "latent_dim": tune.qrandint(2**1, 2**7, 2), # tune.uniform(lambda _: 2 ** np.random.randint(1, 7)),
        "variational_beta": tune.loguniform(1e-3, 1e1),
        "lr": tune.loguniform(1e-4, 1e-1),
        # "weight_decay": tune.loguniform(1e-6, 1e-1),
    }

    bohb_hyperband = HyperBandForBOHB(
        time_attr="training_iteration",
        max_t=max_num_epochs,
        reduction_factor=reduction_factor)

    bohb_search = TuneBOHB(
        # space=config_space,  # If you want to set the space manually
        max_concurrent=8
    )

    stopper = CustomStopper(max_iter=max_num_epochs, val_loss=val_loss)
    result = tune.run(
        Train_Val_VAE,
        name='VAE_Ray_BOHB',
        scheduler=bohb_hyperband,
        search_alg=bohb_search,
        metric="val_loss",
        mode="min",
        verbose=1,
        stop=stopper,
        checkpoint_score_attr="val_loss",
        checkpoint_at_end=True,
        checkpoint_freq=checkpoint_freq,
        keep_checkpoints_num=None,
        resources_per_trial={"cpu": cpus_per_trial, "gpu": gpus_per_trial},
        config=config,
        num_samples=num_samples,
    )

    best_trial = result.get_best_trial("val_loss", "min", "last")
    print(" * Best trial config: {}".format(best_trial.config))
    print(" * Best config is:", result.best_config)
    print(" * Best trial final average validation loss: {}".format(best_trial.last_result["val_loss"]))

    device = "cuda:0" if torch.cuda.is_available() else "cpu"

    # test_acc = test_accuracy(best_trained_model, device)
    # print("Best trial test set accuracy: {}".format(test_acc))
    return device, best_trial.config
`
Here is how I call the main function:
`
device, best_config = main(num_samples=(3**4), max_num_epochs=100, val_loss=(-np.inf), reduction_factor=3,
                                  cpus_per_trial=(1/1), gpus_per_trial = (1/8))
`

The output from `BOHB` is as follow:
`== Status ==
Memory usage on this node: 8.3/125.6 GiB
Using HyperBand: num_stopped=0 total_brackets=1 Round #0: Bracket(Max Size (n)=81, Milestone (r)=1, completed=-3.7%): {PAUSED: 15, PENDING: 1, RUNNING: 1}
Resources requested: 1.0/2 CPUs, 0.12/4 GPUs, 0.0/77.05 GiB heap, 0.0/25.54 GiB objects (0/1.0 accelerator_type:GTX)
Current best trial: 4280a67c with val_loss=407.3801439502571 and parameters={'data_dir': './HPO_DASO/PBT/data', 'latent_dim': 14, 'variational_beta': 8.453041722488535, 'lr': 0.004241011030303081}
Result logdir: ./ray_results/VAE_Ray_BOHB
Number of trials: 17/81 (15 PAUSED, 1 PENDING, 1 RUNNING)
`
As we can see, all 8 jobs are on a single GPU, the other 3 of the total 4 are idle. The CPU resource is also under-use.

**Comment by Shentao-YANG** [comment]
Hi,

Thanks for your reply!

Here is my `tune.Trainable` object:

class Train_Val_VAE(tune.Trainable):
    def setup(self, config):
        print(f"Begin training VAE with config: {config}")
        self.net = VariationalAutoEncoder(latent_dim=config.get("latent_dim", 10),
                                          variational_beta=config.get("variational_beta", 1))
        self.device = "cpu"
        if torch.cuda.is_available():
            self.device = "cuda:0"
            if torch.cuda.device_count() > 1:
                self.net = nn.DataParallel(self.net)
        self.net.to(self.device)
        self.trainloader, self.valloader, _ = \
            get_data_loaders(config.get("data_dir", "./data"), batch_size=int(config.get("batch_size", 128)), val_split=0.2)
        self.optimizer = \
            optim.Adam(params=self.net.parameters(), lr=config.get("lr", 1e-3), weight_decay=config.get("weight_decay", 1e-5))

    def step(self):
        train_VAE(self.net, self.optimizer, self.trainloader, self.device)
        val_loss = test_VAE(self.net, self.valloader, self.device)
        return {"val_loss": val_loss}

    def save_checkpoint(self, checkpoint_dir):
        checkpoint_path = os.path.join(checkpoint_dir, "checkpoint")
        torch.save({
                    "model_state_dict": self.net.state_dict(),
                    "optimizer_state_dict": self.optimizer.state_dict()
                }, checkpoint_path)
        return checkpoint_path

    def load_checkpoint(self, checkpoint_path):
        states = torch.load(checkpoint_path)
        self.net.load_state_dict(states['model_state_dict'])
        self.optimizer.load_state_dict(states['optimizer_state_dict'])

Here is my `Stopper`:

class CustomStopper(tune.Stopper):
        def __init__(self, max_iter=100, val_loss=(-np.inf)):
            self.should_stop = False
            self.max_iter = max_iter
            self.val_loss = val_loss

        def __call__(self, trial_id, result):
            if not self.should_stop and result["val_loss"] < self.val_loss:
                self.should_stop = True
            return self.should_stop or result["training_iteration"] >= self.max_iter

        def stop_all(self):
            return self.should_stop

Here is my `main` function:

def main(num_samples=27, max_num_epochs=200, val_loss=(-np.inf), reduction_factor=3, checkpoint_freq=5,
         cpus_per_trial=(1/1), gpus_per_trial = (1/2)):

    ray.shutdown()
    ray.init(num_cpus=2)

    print(" # check if PytorchTrainble will save/restore correctly before execution")
    validate_save_restore(Train_Val_VAE)
    validate_save_restore(Train_Val_VAE, use_object_store=True)

    data_dir = os.path.abspath("./data")
    get_data_loaders(data_dir, batch_size=64)  # Download data for all trials before starting the run
    config = {
        "data_dir": data_dir,
        "latent_dim": tune.qrandint(2**1, 2**7, 2), # tune.uniform(lambda _: 2 ** np.random.randint(1, 7)),
        "variational_beta": tune.loguniform(1e-3, 1e1),
        "lr": tune.loguniform(1e-4, 1e-1),
        # "weight_decay": tune.loguniform(1e-6, 1e-1),
    }

    bohb_hyperband = HyperBandForBOHB(
        time_attr="training_iteration",
        max_t=max_num_epochs,
        reduction_factor=reduction_factor)

    bohb_search = TuneBOHB(
        # space=config_space,  # If you want to set the space manually
        max_concurrent=8
    )

    stopper = CustomStopper(max_iter=max_num_epochs, val_loss=val_loss)
    result = tune.run(
        Train_Val_VAE,
        name='VAE_Ray_BOHB',
        scheduler=bohb_hyperband,
        search_alg=bohb_search,
        metric="val_loss",
        mode="min",
        verbose=1,
        stop=stopper,
        checkpoint_score_attr="val_loss",
        checkpoint_at_end=True,
        checkpoint_freq=checkpoint_freq,
        keep_checkpoints_num=None,
        resources_per_trial={"cpu": cpus_per_trial, "gpu": gpus_per_trial},
        config=config,
        num_samples=num_samples,
    )

    best_trial = result.get_best_trial("val_loss", "min", "last")
    print(" * Best trial config: {}".format(best_trial.config))
    print(" * Best config is:", result.best_config)
    print(" * Best trial final average validation loss: {}".format(best_trial.last_result["val_loss"]))

    device = "cuda:0" if torch.cuda.is_available() else "cpu"

    # test_acc = test_accuracy(best_trained_model, device)
    # print("Best trial test set accuracy: {}".format(test_acc))
    return device, best_trial.config

Here is how I call the `main`:

device, best_config = main(num_samples=(3**4), max_num_epochs=100, val_loss=(-np.inf), reduction_factor=3,
                                  cpus_per_trial=(1/1), gpus_per_trial = (1/8))

Here is the `output`:

== Status ==
Memory usage on this node: 8.3/125.6 GiB
Using HyperBand: num_stopped=0 total_brackets=1 Round #0: Bracket(Max Size (n)=81, Milestone (r)=1, completed=-3.7%): {PAUSED: 15, PENDING: 1, RUNNING: 1}
Resources requested: 1.0/2 CPUs, 0.12/4 GPUs, 0.0/77.05 GiB heap, 0.0/25.54 GiB objects (0/1.0 accelerator_type:GTX)
Current best trial: 4280a67c with val_loss=407.3801439502571 and parameters={'data_dir': '//HPO_DASO/PBT/data', 'latent_dim': 14, 'variational_beta': 8.453041722488535, 'lr': 0.004241011030303081}
Result logdir: /ray_results/VAE_Ray_BOHB
Number of trials: 17/81 (15 PAUSED, 1 PENDING, 1 RUNNING)

As we can see, all 8 jobs are on a single GPU and the remaining 3 out of total 4 GPUs are idle. Also, the CPU is under-utilized.
BTW, is there a way to check the GPU memory during ray runtime ?

**Comment by amogkam** [comment]
It looks like your cluster only has 2 CPUs available, and since each trial requires 1 CPU, at most only 2 trials will ever be run in parallel. As a result all 4 GPUs will never be used in your current setup- the available CPU is what's limiting the parallelism. To fix this you can either increase the number of CPUs in the cluster, decrease the `cpus_per_trial` value to a fractional value, or increase the `gpus_per_trial` value so that even if you have a max of 2 trials running in parallel all 4 GPUs will be used.

**Comment by Shentao-YANG** [comment]
Hi, 

Thanks for reply.

I decrease the `cpus_per_trial` value on to a fractional value $2/32$ but the GPUs still seem not fully utilized. The code and output are as follows.
`
device, best_config = main(num_samples=(3**4), max_num_epochs=100, val_loss=(-np.inf), reduction_factor=3,
                                  cpus_per_trial=(2/32), gpus_per_trial = (4/32))
`

`
== Status ==
Memory usage on this node: 10.4/125.6 GiB
Using HyperBand: num_stopped=0 total_brackets=1 Round #0: Bracket(Max Size (n)=81, Milestone (r)=1, completed=-13.6%): {PAUSED: 55, RUNNING: 1}
Resources requested: 0.06/2 CPUs, 0.12/4 GPUs, 0.0/77.88 GiB heap, 0.0/25.78 GiB objects (0/1.0 accelerator_type:GTX)
Number of trials: 56/81 (55 PAUSED, 1 RUNNING)
`
I also change the `device` in Pytorch from `cuda:0` to `cuda` but nothing changed.

**Comment by altosaar** [comment]
Ran into a similar issue. Torch could see the GPU but ray trials kept erroring out with 'CUDA not available' - very tricky to debug.

Changing the way the device was specified from `device = torch.device(0)` to `device = "cuda:0"` as in https://docs.ray.io/en/latest/tune/tutorials/tune-pytorch-cifar.html#adding-multi-gpu-support-with-dataparallel fixed it.

**Comment by richardliaw** [comment]
@altosaar that's great to hear! I'll be sure to refer others to this particular solution in the future.

Thanks for sharing!

**Comment by richardliaw** [comment]
@Shentao-YANG I would suggest turning off Hyperband for now to validate the behavior properly.

Closing this issue because I think the original issue was addressed, and the behavior @Shentao-YANG is seeing now is just of Hyperband.

**Comment by ferhatmelih** [comment]
Hi, Thanks for the solutions above but some of the links are old. I want to share my solution based on this documentation: https://docs.ray.io/en/latest/tune/tutorials/tune-resources.html 

```python
def test_pytorch(config):
    print('is cuda avaiable ',torch.cuda.is_available())
    tune.report({'metric':0})

tuner = tune.Tuner(tune.with_resources(test_pytorch, {"gpu": 1}), 
   tune_config=tune.TuneConfig(num_samples=10))
results = tuner.fit()
```
